File: okcanvas-agent-runtime/clients/cli/src/sse.ts
import { CliError } from "./config.js";
import type { RunEvent } from "./types.js";

export async function* parseSseJson(stream: ReadableStream<Uint8Array>): AsyncGenerator<RunEvent> {
  const reader = stream.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  let dataLines: string[] = [];

  const dispatch = (): RunEvent | null => {
    if (dataLines.length === 0) return null;
    const raw = dataLines.join("\n");
    dataLines = [];
    let payload: unknown;
    try {
      payload = JSON.parse(raw);
    } catch {
      throw new CliError("CLI_SSE_DATA_INVALID", "Control API persisted SSE returned invalid JSON");
    }
    if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
      throw new CliError("CLI_SSE_DATA_INVALID", "Control API persisted SSE returned a non-object payload");
    }
    return payload as RunEvent;
  };

  while (true) {
    const { done, value } = await reader.read();
    buffer += decoder.decode(value, { stream: !done });
    const lines = buffer.split(/\r?\n/u);
    buffer = done ? "" : (lines.pop() ?? "");
    for (const line of lines) {
      if (line === "") {
        const event = dispatch();
        if (event) yield event;
      } else if (line.startsWith("data:")) {
        dataLines.push(line.slice(5).replace(/^ /u, ""));
      }
    }
    if (done) break;
  }
  if (buffer) {
    for (const line of buffer.split(/\r?\n/u)) {
      if (line.startsWith("data:")) dataLines.push(line.slice(5).replace(/^ /u, ""));
    }
  }
  const event = dispatch();
  if (event) yield event;
}

File: okcanvas-agent-runtime/okcanvas_agent_clients/tui/sse.py
from __future__ import annotations

import json
from collections.abc import Iterable, Iterator
from dataclasses import dataclass
from typing import Any

from okcanvas_agent_clients.tui.config import TUIClientError


@dataclass(frozen=True)
class SSEMessage:
    event_id: str | None
    event_type: str | None
    data: str


def parse_sse_lines(lines: Iterable[str]) -> Iterator[SSEMessage]:
    event_id: str | None = None
    event_type: str | None = None
    data_lines: list[str] = []

    def dispatch() -> SSEMessage | None:
        nonlocal event_id, event_type, data_lines
        if not data_lines:
            event_id = None
            event_type = None
            return None
        message = SSEMessage(
            event_id=event_id,
            event_type=event_type,
            data="\n".join(data_lines),
        )
        event_id = None
        event_type = None
        data_lines = []
        return message

    for raw_line in lines:
        line = raw_line.rstrip("\r\n")
        if line == "":
            message = dispatch()
            if message is not None:
                yield message
            continue
        if line.startswith(":"):
            continue
        field, separator, value = line.partition(":")
        if separator and value.startswith(" "):
            value = value[1:]
        if field == "id":
            event_id = value
        elif field == "event":
            event_type = value
        elif field == "data":
            data_lines.append(value)

    message = dispatch()
    if message is not None:
        yield message


def parse_sse_json(lines: Iterable[str]) -> Iterator[dict[str, Any]]:
    for message in parse_sse_lines(lines):
        try:
            payload = json.loads(message.data)
        except json.JSONDecodeError as exc:
            raise TUIClientError(
                "TUI_SSE_DATA_INVALID",
                "Control API persisted SSE returned invalid JSON",
            ) from exc
        if not isinstance(payload, dict):
            raise TUIClientError(
                "TUI_SSE_DATA_INVALID",
                "Control API persisted SSE returned a non-object payload",
            )
        yield payload
